src/app/modules/twenty-four-hour-time/twenty-four-hour-time.component.ts

```diff
@@ -0,0 +1,309 @@
+      event.target.value = formattedInput;
+      this.control.setValue(formattedInput);
+      this.knownGoodTimeValue = formattedInput;
+      this.announceCurrentTime();
+      this.timeHours = hours;
+      this.timeMinutes = minutes;
+      return;
+    }
+    this.addValidationError();
+    event.target.value = this.knownGoodTimeValue ?? '';
+    this.control.setValue(this.knownGoodTimeValue);
+  }
+
+  // ensure that the user can only enter numbers and a colon
+  // this function is called for every keystroke
+  validateInput(event) {
+    // close the time picker if it's open
+    this.timePickerVisible = false;
+
+    const input = event.target.value;
+    const permittedCharacters = /^[0-9:]{0,5}$/;
+    const forbiddenCharacters = /[^[0-9:]/g;
+    this.removeValidationError();
+    if (!permittedCharacters.test(input)) {
+      event.target.value = input.replace(forbiddenCharacters, '');
+    }
+
+    clearTimeout(this.inputDebounceTimeout);
+    // attempt to format and set time if we haven't gotten any more inputs in the past second
+    // don't attempt to validate if user hasn't entered at least 3 characters
+    if (input.replace(':', '').length >= 3) {
+      this.inputDebounceTimeout = setTimeout(() => {
+        this.formatAndSetTime(event);
+      }, 1000);
+    }
+  }
+
+  openTimePicker(event: Event) {
+    if (!this.disabled) {
+      this.removeValidationError();
+      this.addFocusStyling();
+      event.stopPropagation();
+      event.preventDefault();
+
+      // if we open the time picker before any time is set, default to midday
+      if (!this.control.value) {
+        this.control.setValue('12:00');
+        this.knownGoodTimeValue = '12:00';
+        this.timeHours = '12';
+        this.timeMinutes = '00';
+      }
+      this.calculateSurroundingHours();
+      this.calculateSurroundingMinutes();
+      this.timePickerVisible = true;
+      setTimeout(() => this.focusOnCloseButton(event), 5);
+    }
+  }
+
+  focusLost(event: any) {
+    // focusout fires when the time picker loses focus
+    // check that the new element gaining focus doesn't also live inside the time picker
+    // if it doesn't, we can safely close the time picker
+    // ignore these clicks and only close the timepicker when we're clicking outside the page
+    const timePickerClasses = ['bumper-icon', 'time-picker-body', 'time-picker-label', 'close-image'];
+    if (timePickerClasses.every((tpc) => !event.relatedTarget?.classList?.contains(tpc))) {
+      this.removeFocusStyling();
+      this.closeTimePicker(event);
+    }
+  }
+
+  closeTimePicker(event: Event) {
+    event.stopPropagation();
+    event.preventDefault();
+    this.timePickerVisible = false;
+  }
+
+  stepTime(unit: 'hours' | 'minutes', direction: 'increment' | 'decrement') {
+    if (unit === 'hours') {
+      // calculate hour step
+      const currentHour = Number(this.timeHours);
+      let newHour;
+      if (direction === 'increment' && currentHour < 23) {
```

Comment:
I really prefer a newline above some block of core, or a group of related operations. Or at the very least after variable declarations, or just above some conditional block.